# Resolution: Transfer Response Model

Consequent Reversibility was not included in the final escape models because, in this subgroup structure, it is partially encoded by Goal_Type and produced unstable estimates when entered together with Goal_Type.

Import Libraries

In [34]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Overall Subgroup
This subset of data explores observations when _Focus_ = I and _Focus_= They.
1. Create resolved_transfer
2. Counts and Proportions
3. Logistic Models
    - RT1: Agency
    - RT2: Goal Type
    - RT3: Additive Model
    - RT4: Interactive Model 
4. Model Comparison
5. Summary Table

Read Data

In [35]:
subgroup_theoretical = pd.read_csv("../../data/processed/subgroup_theoretical.csv")

1. Create resolved_transfer

Binary Outcome: include L1_transfer responses from escape_L2 observations.

In [36]:
subgroup_theoretical["escape_L2"] = (
    subgroup_theoretical["Response_Full"] != "L2_other"
).astype(int)

In [37]:
escapees = subgroup_theoretical[subgroup_theoretical["escape_L2"] == 1].copy()

In [38]:
escapees["resolved_transfer"] = (escapees["Response_Full"] == "L1_transfer").astype(int)

2. Counts and Proportions

In [39]:
resolved_transfer_counts = pd.crosstab(
    escapees["resolved_transfer"],
    escapees["Response_Full"])

resolved_transfer_counts

Response_Full,L1_transfer,correct,missing_response
resolved_transfer,,,
0,0,61,3
1,96,0,0


In [40]:
resolved_transfer_props = escapees["resolved_transfer"].value_counts(normalize = True)
resolved_transfer_props

resolved_transfer
1    0.6
0    0.4
Name: proportion, dtype: float64

From 160 observations, 60% of responses that had escaped *L2_other* options were resolved by restoring to L1_transfer. 

Counts and Proportions by Condition:

In [41]:
counts_condition = pd.crosstab(
    [escapees["Goal_Type"], escapees["Agent"]],
    escapees["resolved_transfer"],
    margins = True)

counts_condition

resolved_transfer         0   1  All
Goal_Type         Agent             
goal_frequent     0      15  13   28
                  1      10  12   22
goal_non_frequent 0      12  21   33
                  1       7  16   23
no_goal           0      11  23   34
                  1       9  11   20
All                      64  96  160

In [42]:
props_condition = pd.crosstab(
    [escapees["Goal_Type"], escapees["Agent"]],
    escapees["resolved_transfer"],
    normalize= "index")

props_condition

resolved_transfer               0         1
Goal_Type         Agent                    
goal_frequent     0      0.535714  0.464286
                  1      0.454545  0.545455
goal_non_frequent 0      0.363636  0.636364
                  1      0.304348  0.695652
no_goal           0      0.323529  0.676471
                  1      0.450000  0.550000

COMMENT GENERAL PROPORTIONS

3. Logistic Models

Import Libraries

In [43]:
import statsmodels.formula.api as smf
from scipy.stats import chi2

In [44]:
escapees["resolved_transfer"].value_counts()

resolved_transfer
1    96
0    64
Name: count, dtype: int64

**Model RT1**

resolved_transfer ~ Agent

In [45]:
RT1 = smf.logit(
    "resolved_transfer ~ Agent",
    data=escapees
    ).fit()

print(RT1.summary())

Optimization terminated successfully.
         Current function value: 0.673012
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                  160
Model:                          Logit   Df Residuals:                      158
Method:                           MLE   Df Model:                            1
Date:                Sat, 27 Jun 2026   Pseudo R-squ.:               1.359e-10
Time:                        00:08:29   Log-Likelihood:                -107.68
converged:                       True   LL-Null:                       -107.68
Covariance Type:            nonrobust   LLR p-value:                    0.9999
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.4055      0.209      1.936      0.053      -0.005       0.816
Agent      -9.339e-16      0.

**Model RT2**

resolved_transfer ~ Goal_Type

In [46]:
RT2 = smf.logit(
    "resolved_transfer ~ Goal_Type",
    data= escapees
    ).fit()

print(RT2.summary())

Optimization terminated successfully.
         Current function value: 0.663269
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                  160
Model:                          Logit   Df Residuals:                      157
Method:                           MLE   Df Model:                            2
Date:                Sat, 27 Jun 2026   Pseudo R-squ.:                 0.01448
Time:                        00:08:29   Log-Likelihood:                -106.12
converged:                       True   LL-Null:                       -107.68
Covariance Type:            nonrobust   LLR p-value:                    0.2104
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                      -1.852e-16      0.283  -6.55e-16     

**Model RT3**

resolved_transfer ~ Agent + Goal_Type

In [47]:
RT3 = smf.logit(
    "resolved_transfer ~ Agent + Goal_Type",
    data= escapees
    ).fit()

print(RT3.summary())

Optimization terminated successfully.
         Current function value: 0.663253
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                  160
Model:                          Logit   Df Residuals:                      156
Method:                           MLE   Df Model:                            3
Date:                Sat, 27 Jun 2026   Pseudo R-squ.:                 0.01450
Time:                        00:08:29   Log-Likelihood:                -106.12
converged:                       True   LL-Null:                       -107.68
Covariance Type:            nonrobust   LLR p-value:                    0.3731
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                         -0.0103      0.318     -0.032     

**Model RT4**

resolved_transfer ~ Agent * Goal_Type

In [48]:
RT4 = smf.logit(
    "resolved_transfer ~ Agent * Goal_Type",
    data=escapees
    ).fit()

print(RT4.summary())

Optimization terminated successfully.
         Current function value: 0.658907
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                  160
Model:                          Logit   Df Residuals:                      154
Method:                           MLE   Df Model:                            5
Date:                Sat, 27 Jun 2026   Pseudo R-squ.:                 0.02096
Time:                        00:08:29   Log-Likelihood:                -105.43
converged:                       True   LL-Null:                       -107.68
Covariance Type:            nonrobust   LLR p-value:                    0.4781
                                           coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                               -0.1431      0.3

4. Model Comparisons

In [49]:
from scipy.stats import chi2

def lr_test(model_small, model_large):

    LR = 2 * (model_large.llf - model_small.llf)

    df = model_large.df_model - model_small.df_model

    p = chi2.sf(LR, df)

    return pd.Series({
        "LR": LR,
        "df": df,
        "p_value": p
    })

In [50]:
lr_test(RT1, RT3)

LR         3.122680
df         2.000000
p_value    0.209855
dtype: float64

In [51]:
lr_test(RT1, RT4)

LR         4.513476
df         4.000000
p_value    0.340953
dtype: float64

5. Summary Table

model_name = ["RT1", "RT2", "RT2", "RT3", "RT3", "RT3", "RT$", "RT4", "RT4", "RT4", "RT4"]

outcome = ["resolved_transfer","resolved_transfer", "resolved_transfer", "resolved_transfer", "resolved_transfer", "resolved_transfer", "resolved_transfer", "resolved_transfer", "resolved_transfer", "resolved_transfer", "resolved_transfer"]

subgroup = ["Overall", "Overall", "Overall", "Overall", "Overall", "Overall", "Overall", "Overall", "Overall", "Overall", "Overall"]

formula = ["resolved_transfer ~ Agent", "resolved_transfer ~ Goal_Type", "resolved_transfer ~ Goal_Type", "resolved_transfer ~ Agent + Goal_Type", "resolved_transfer ~ Agent + Goal_Type", "resolved_transfer ~ Agent + Goal_Type", "resolved_transfer ~ Agent * Goal_Type", "resolved_transfer ~ Agent * Goal_Type", "resolved_transfer ~ Agent * Goal_Type", "resolved_transfer ~ Agent * Goal_Type", "resolved_transfer ~ Agent * Goal_Type"]

predictor= ["Agent", "goal_non_frequent", "no_goal", "goal_non_frequent", "no_goal", "Agent",  "goal_non_frequent", "no_goal", "Agent", "Agent:goal_non_frequent", "Agent:no_goal"]

coef = [1.1086, 0.5833, 0.3646, 0.6038, 0.3785, 1.1182, 0.6061, 0.7538, 1.7047, 0.1315, -1.5422]

coef_p = [0.013, 0.193, 0.396, 0.185, 0.387, 0.013, 0.224, 0.141, 0.035, 0.923, 0.143]

LLR_p = [0.007642, 0.4060, 0.4060, 0.02954, 0.02954, 0.02954, 0.03590, 0.03590, 0.03590, 0.03590, 0.03590]

N= [198, 198, 198, 198, 198, 198, 198, 198, 198, 198, 198]

Takeaway = []

correct_summary_overall = pd. DataFrame({
    "Model": model_name,
    "Outcome": outcome,
    "Subgroup": subgroup,
    "Formula": formula, 
    "Predictor": predictor,
    "Coef": coef,
    "Coef_p": coef_p,
    "Model_LLR_p": LLR_p,
    "N": N
})

correct_summary_overall

## Actors Subgroup
This subset of data explores observations only when _Focus_ = I.
1. Create resolved_transfer
2. Counts and Proportions
3. Logistic Models
    - RT1_A: Agency
    - RT2_A: Goal Type
    - RT3_A: Additive Model
    - RT4_A: Interactive Model 
4. Model Comparison
5. Summary Table

Read Data

In [52]:
subgroup_actors = subgroup_theoretical[subgroup_theoretical["Focus"] == "I"].copy()

1. Create resolved_transfer

Binary Outcome: include transfer responses from escape_L2 observations within actors.

In [53]:
subgroup_actors["escape_L2"] = (
    subgroup_actors["Response_Full"] != "L2_other"
).astype(int)

In [54]:
escapees_actors = subgroup_actors[subgroup_actors["escape_L2"] == 1].copy()

In [55]:
escapees_actors["resolved_transfer"] = (escapees_actors["Response_Full"] == "L1_transfer").astype(int)

2. Counts and Proportions

In [56]:
resolved_transfer_counts = pd.crosstab(
    escapees_actors["resolved_transfer"],
    escapees_actors["Response_Full"])

resolved_transfer_counts

Response_Full,L1_transfer,correct,missing_response
resolved_transfer,,,
0,0,40,3
1,53,0,0


In [57]:
resolved_transfer_props = escapees_actors["resolved_transfer"].value_counts(normalize = True)
resolved_transfer_props

resolved_transfer
1    0.552083
0    0.447917
Name: proportion, dtype: float64

From 96 actors' observations (from the original 160), 55.2% of responses that managed to escape *L2_other* options were resolved by restoring to *L1_transfer*.

Counts and Proportions by Condition:

In [58]:
counts_condition = pd.crosstab(
    [escapees_actors["Goal_Type"], escapees_actors["Agent"]],
    escapees_actors["resolved_transfer"],
    margins = True)

counts_condition

resolved_transfer         0   1  All
Goal_Type         Agent             
goal_frequent     0      10   6   16
                  1       7   7   14
goal_non_frequent 0       9   7   16
                  1       4  11   15
no_goal           0       8  13   21
                  1       5   9   14
All                      43  53   96

In [59]:
props_condition = pd.crosstab(
    [escapees_actors["Goal_Type"], escapees_actors["Agent"]],
    escapees_actors["resolved_transfer"],
    normalize= "index")

props_condition

resolved_transfer               0         1
Goal_Type         Agent                    
goal_frequent     0      0.625000  0.375000
                  1      0.500000  0.500000
goal_non_frequent 0      0.562500  0.437500
                  1      0.266667  0.733333
no_goal           0      0.380952  0.619048
                  1      0.357143  0.642857

ADD COMMENT ON GENERAL PROPORTIONS 

3. Logistic Models

In [60]:
escapees_actors["resolved_transfer"].value_counts()

resolved_transfer
1    53
0    43
Name: count, dtype: int64

**RT1_A Model** 

resolved_transfer ~ Agent

In [61]:
RT1_A = smf.logit(
    "resolved_transfer ~ Agent",
    data=escapees_actors
    ).fit()

print(RT1_A.summary())

Optimization terminated successfully.
         Current function value: 0.678229
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                   96
Model:                          Logit   Df Residuals:                       94
Method:                           MLE   Df Model:                            1
Date:                Sat, 27 Jun 2026   Pseudo R-squ.:                 0.01379
Time:                        00:08:29   Log-Likelihood:                -65.110
converged:                       True   LL-Null:                       -66.020
Covariance Type:            nonrobust   LLR p-value:                    0.1772
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.0377      0.275     -0.137      0.891      -0.576       0.501
Agent          0.5610      0.

**RT2_A Model** 

resolved_transfer ~  Goal_Type

In [62]:
RT2_A = smf.logit(
    "resolved_transfer ~ Goal_Type",
    data=escapees_actors
    ).fit()

print(RT2_A.summary())

Optimization terminated successfully.
         Current function value: 0.673952
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                   96
Model:                          Logit   Df Residuals:                       93
Method:                           MLE   Df Model:                            2
Date:                Sat, 27 Jun 2026   Pseudo R-squ.:                 0.02001
Time:                        00:08:29   Log-Likelihood:                -64.699
converged:                       True   LL-Null:                       -66.020
Covariance Type:            nonrobust   LLR p-value:                    0.2669
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                         -0.2683      0.368     -0.728     

**RT3_A Model**

resolved_transfer ~  Agent + Goal_Type

In [63]:
RT3_A = smf.logit(
    "resolved_transfer ~ Agent + Goal_Type",
    data=escapees_actors
    ).fit()

print(RT3_A.summary())

Optimization terminated successfully.
         Current function value: 0.663120
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                   96
Model:                          Logit   Df Residuals:                       92
Method:                           MLE   Df Model:                            3
Date:                Sat, 27 Jun 2026   Pseudo R-squ.:                 0.03576
Time:                        00:08:29   Log-Likelihood:                -63.659
converged:                       True   LL-Null:                       -66.020
Covariance Type:            nonrobust   LLR p-value:                    0.1933
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                         -0.5595      0.427     -1.311     

**RT4_A Model**

resolved_transfer ~ Agent * Goal_Type

In [64]:
RT4_A = smf.logit(
    "resolved_transfer ~ Agent * Goal_Type",
    data= escapees_actors
    ).fit()

print(RT4_A.summary())

Optimization terminated successfully.
         Current function value: 0.656589
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                   96
Model:                          Logit   Df Residuals:                       90
Method:                           MLE   Df Model:                            5
Date:                Sat, 27 Jun 2026   Pseudo R-squ.:                 0.04526
Time:                        00:08:29   Log-Likelihood:                -63.033
converged:                       True   LL-Null:                       -66.020
Covariance Type:            nonrobust   LLR p-value:                    0.3086
                                           coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                               -0.5108      0.5

4. Model comparisons

In [65]:
lr_test(RT1_A, RT3_A)

LR         2.900943
df         2.000000
p_value    0.234460
dtype: float64

In [66]:
lr_test(RT1_A, RT4_A)

LR         4.154880
df         4.000000
p_value    0.385451
dtype: float64

5. Summary Table

correct_summary_actors

model_name = ["RT1_A", "RT2_A", "RT2_A", "RT3_A", "RT3_A", "RT3_A", "RCT_A", "RT4_A", "RT4_A", "RT4_A", "RT4_A"]

outcome = ["resolved_transfer", "resolved_transfer", "resolved_transfer", "resolved_transfer", "resolved_transfer", "resolved_transfer", "resolved_transfer", "resolved_transfer", "resolved_transfer", "resolved_transfer", "resolved_transfer"]

subgroup = ["Actors", "Actors", "Actors", "Actors", "Actors", "Actors", "Actors", "Actors", "Actors", "Actors", "Actors"]

formula = ["resolved_transfer ~ Agent", "resolved_transfer ~ Goal_Type", "resolved_transfer ~ Goal_Type", "resolved_transfer ~ Agent + Goal_Type", "resolved_transfer ~ Agent + Goal_Type", "resolved_transfer ~ Agent + Goal_Type", "resolved_transfer ~ Agent * Goal_Type", "resolved_transfer ~ Agent * Goal_Type", "resolved_transfer ~ Agent * Goal_Type", "resolved_transfer ~ Agent * Goal_Type", "resolved_transfer ~ Agent * Goal_Type"]

predictor= ["Agent", "goal_non_frequent", "no_goal", "goal_non_frequent", "no_goal", "Agent",  "goal_non_frequent", "no_goal", "Agent", "Agent:goal_non_frequent", "Agent:no_goal"]

coef = [1.1259, 0.1382, 0.8473, 0.1448, 0.8790, 1.1455, -3.881e-16, 1.2528, 1.2528, 0.7621, -1.2528]

coef_p = [0.038, 0.793, 0.159, 0.788, 0.151, 0.036, 1.000, 0.097, 0.150, 0.591, 0.338]

LLR_p = [0.02657, 0.3143, 0.3143, 0.06234, 0.06234, 0.06234, 0.09726, 0.09726, 0.09726, 0.09726, 0.09726]

N = [120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120]

Takeaway = []

correct_summary_actors = pd. DataFrame({
    "Model": model_name,
    "Outcome": outcome,
    "Subgroup": subgroup,
    "Formula": formula, 
    "Predictor": predictor,
    "Coef": coef,
    "Coef_p": coef_p,
    "Model_LLR_p": LLR_p,
    "N": N
})

correct_summary_actors